Contracts/Presale/PresaleContract.sol

```diff
@@ -23,88 +23,87 @@ contract PresaleContract is WithdrawContract {
+        amountsLeft.push(THIRD_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
     }
     
     mapping (address => Balance) usersBalance;
     Transaction[] transactions;
     
-    function buyTokens() public payable {
+    function () external payable {
         require(msg.value >= MINIMUM_PAYMENT_VALUE);
-        uint deposit = msg.value;
+        uint totalDeposit = msg.value;
         
-        uint8 index = 0;
-        uint profit = 0;
+        uint8 stage = 0;
+        uint currentInvestment = 0;
         
         uint totalTokens = 0;
         
-        while (deposit > 0 && index < 3) {
-            uint ethers = 0;
-            uint maxAmount = deposit * MINIMUM_TOKEN_VALUE_MULTIPLIER * prices[index] / 1 ether;
-            if (amounts[index] >= maxAmount) {
-                amounts[index] -= maxAmount;
-                ethers = deposit;
-                deposit = 0;
+        while (totalDeposit > 0 && stage < 3) {
+            uint currentStageDeposit = 0;
+            
+            uint maxTokensAmountForCurrentStage = totalDeposit * amountBoughtInOneEth[stage];
+            //checking free tokens amount for current stage
+            if (amountsLeft[stage] >= maxTokensAmountForCurrentStage) {
+                amountsLeft[stage] -= maxTokensAmountForCurrentStage;
+                currentStageDeposit = totalDeposit;
+                totalDeposit = 0;
             } else {
-                maxAmount = amounts[index];
-                amounts[index] = 0;
-                ethers = maxAmount * 1 ether / MINIMUM_TOKEN_VALUE_MULTIPLIER / prices[index];
-                deposit -= ethers;
+                maxTokensAmountForCurrentStage = amountsLeft[stage];
+                amountsLeft[stage] = 0;
+                currentStageDeposit = maxTokensAmountForCurrentStage / amountBoughtInOneEth[stage];
+                totalDeposit -= currentStageDeposit;
                 
             }
             //save transaction
-            _addTransaction(msg.sender, ethers, prices[index], maxAmount);
+            _addTransaction(msg.sender, currentStageDeposit, amountBoughtInOneEth[stage], maxTokensAmountForCurrentStage);
             //calculations
-            totalTokens += maxAmount;
-            profit += ethers;
+            totalTokens += maxTokensAmountForCurrentStage;
+            currentInvestment += currentStageDeposit;
             //
-            index++;
+            stage++;
         }
         
-        _updateData(msg.sender, profit, totalTokens);
+        _updateUserBalance(msg.sender, currentInvestment, totalTokens);
         //return money
-        if (deposit > 0) {
-            withdraw(msg.sender, deposit);
+        if (totalDeposit > 0) {
+            withdraw(msg.sender, totalDeposit);
         }
         //auto-transfer to our wallet
-        if (profit > 0) {
-            withdraw(mainVallet, profit);
-            // mainVallet.transfer(profit);
+        if (currentInvestment > 0) {

```

Comment:
Put these lines (current investment transfer) before the return to user - just in case ..